### PR TITLE
qemu_vm: enhancement of the reboot function

### PR DESCRIPTION
**Pull request by luckyh**
This patch makes checking VM's shutdown event by listening the "RESET"
event from QMP (if supported) as the default way in the reboot
function.

ID: 1439059

### CODE REVIEW

**Comment by pingl2016** [comment]
09:55:50 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_framework-49.0-py2.7.egg/avocado/core/test.py:596
09:55:50 ERROR| Traceback (most recent call last):
09:55:50 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-49.0-py2.7.egg/avocado_vt/test.py", line 271, in runTest
09:55:50 ERROR|     raise self.__status
09:55:50 ERROR| ShellCmdError: Shell command failed: u'python D:\\win_driver_install.py --verify_driver --driver_path E:\\balloon\\2k8\\amd64 --driver_name balloon --device_name "VirtIO Balloon Driver" --vol_utils D:'    (status: 1,    output: u'\n')
09:55:50 ERROR|
09:55:50 DEBUG| Local variables:
09:55:50 DEBUG|  -> self <class 'avocado_vt.test.VirtTest'>: 2-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2008.x86_64.sp2.io-github-autotest-qemu.single_driver_install.with_balloon
09:55:50 ERROR| Traceback (most recent call last):

09:55:50 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_framework-49.0-py2.7.egg/avocado/core/test.py", line 667, in _run_avocado
    raise test_exception

09:55:50 ERROR| ShellCmdError: Shell command failed: u'python D:\\win_driver_install.py --verify_driver --driver_path E:\\balloon\\2k8\\amd64 --driver_name balloon --device_name "VirtIO Balloon Driver" --vol_utils D:'    (status: 1,    output: u'\n')

09:55:50 ERROR| ERROR 2-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Win2008.x86_64.sp2.io-github-autotest-qemu.single_driver_install.with_balloon -> ShellCmdError: Shell command failed: u'python D:\\win_driver_install.py --verify_driver --driver_path E:\\balloon\\2k8\\amd64 --driver_name balloon --device_name "VirtIO Balloon Driver" --vol_utils D:'    (status: 1,    output: u'\n')
09:55:50 INFO |

**Comment by pingl2016** [comment]
Acked-by: Ping Li \<pingl@redhat.com\>
acceptance and tree signoff test passed with the patch

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -4015,57 +4014,49 @@ def _go_down(session, timeout):
...
+            _check_go_down = partial(_go_down, session, timeout)
         elif method == "system_reset":
-            self.system_reset()
+            _reboot = self.system_reset
+            _check_go_down = partial(bool, True)
         else:
             raise virt_vm.VMRebootError("Unknown reboot method: %s" % method)
+
+        if isinstance(self.monitor, qemu_monitor.QMPMonitor):
+            _check_go_down = partial(self.monitor.get_event, "RESET")
```
Comment (CongLi) [inline virttest/qemu_vm.py]: clear 'RESET' events before trigger it ?

Comment (luckyh) [inline virttest/qemu_vm.py]: The `_check_go_down` function will be called at line 4049, and the next line will clear `RESET` events, so I think it should be ok.

Comment (luckyh) [inline virttest/qemu_vm.py]: The `_check_go_down` function will be called at line 4049, and the next line will clear `RESET` events, so I think it should be ok.

Comment (CongLi) [inline virttest/qemu_vm.py]: I was just thinking about the kernel crash and guest reboot automatically.

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**